## Profitable App Profiles for the App Store and Google Play Markets

This project aims to analyse IOS and Android app data to help developers gain a better understanding of what type of apps are likely to attract more users on both Google Play and the App Store.

---

In [1]:
from csv import reader

# function to read dataset
def read_dataset(file, header=False):
    opened_file = open(file, encoding='utf8')
    read_file = reader(opened_file)
    dataset = list(read_file)
    if header:
        header = dataset[0]
        dataset = dataset[1:]
        return dataset, header
    else:
        return dataset

# read datasets and store to respective variables
ios_data, ios_header = read_dataset('data/AppleStore.csv', header=True)
android_data, android_header = read_dataset('data/googleplaystore.csv', header=True)

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
# explore app store data
print('IOS app data:\n')
explore_data(ios_data, 1, 4, True)

# explore android app data
print('\n\nAndroid app data:\n')
explore_data(android_data, 1, 4, True)

IOS app data:

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 7197
Number of columns: 16


Android app data:

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 

---
From looking at the columns below, I think that:

for the IOS app data (available [here](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps?select=AppleStore.csv)). The most useful features will be:

| column name      | index |
| ---------------- | ----- |
| size_bytes       | 0     |
| price            | 4     |
| rating_count_tot | 5     |
| user_rating      | 7     |
| cont_rating      | 10    |
| prime_genre      | 11    |


for the android app data (available [here](https://www.kaggle.com/lava18/google-play-store-apps)). The most useful features will be:

| column name    | index |
| -------------- | ----- |
| category       | 1     |
| rating         | 2     |
| reviews        | 3     |
| size           | 4     |
| installs       | 5     |
| type           | 6     |
| price          | 7     |
| content_rating | 8     |
| genres         | 9     |

---

## Delete samples with missing values

The Google Play data set has a dedicated discussion section, and we can see that one of the discussions (available [here](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015)) outlines an error for row 10472. Let's print this row and compare it against the header and another row that is correct.

---

In [3]:
print(android_data[10472])  # incorrect row
print('\n')
print(android_header)  # header
print('\n')
print(android_data[0])      # correct row

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


---

The row 10472 corresponds to the app 'Life Made WI-Fi Touchscreen Photo Frame', and we can see that the rating is 19. This is clearly off because the maximum rating for a Google Play app is 5 (as mentioned in the discussions section, this problem is caused by a missing value in the 'Category' column). As a consequence, we'll delete this row.

---

In [4]:
# delete row with missing value for android data
del android_data[10472]

---

## Remove duplicate samples

As shown below, the android dataset contains duplicate entries. It is important these duplicates are removed before analysis as we don't want to analyse the same app more than once. 

From the output below, we can see that the duplicate entries for 'Instagram' only differ for their number of reviews. This is because the data samples have been taken at different times. 

---

In [5]:
for i in android_data:
    name = i[0]
    if name == 'Instagram':
        print(i)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


---

The code below shows the total number of dupliate values. The strategy for removing these duplicate values will be to only keep the most recent entry, which is indicated through the sample with the highest number of reviews for each duplicate app. 

---

In [6]:
# show total number of duplicate values
duplicate_apps = []
unique_apps = []

for i in android_data:
    name = i[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate values:', len(duplicate_apps))

Number of duplicate values: 1181


In [7]:
# create a dictionary where the key is a unique app name and the corresponding dictionary value is the highest number of reviews
reviews_max = {}
for i in android_data:
    name = i[0]
    n_reviews = float(i[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

# use created dictionary to remove duplicate rows
android_clean = []
already_added = []
for i in android_data:
    name = i[0]
    n_reviews = float(i[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(i)
        already_added.append(name)

---

## Handling non-english apps

If you explore the data sets enough, you'll notice the names of some of the apps suggest they are not directed toward an English-speaking audience.

We're not interested in keeping these kind of apps, so we'll remove them. One way to go about this is to remove each app whose name contains a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;, etc.), and other symbols (+, *, /, etc.).

All these characters that are specific to English texts are encoded using the ASCII standard. Each ASCII character has a corresponding number between 0 and 127 associated with it, and we can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters.

The function below, uses the built-in ord() function to find out the corresponding encoding number of each character. To minimize the impact of data loss, we'll only remove an app if its name has more than three non-ASCII characters.

---

In [8]:
# function for detecting string with non-english characters
def english_char(string):
    n_char = 0
    for i in string:
        if ord(i) > 127:
            n_char += 1
    
    if n_char > 3:
        return False
    else:
        return True

# testing function
print('String: \'Instagram\', is in english:', english_char('Instagram'))
print('String: \'爱奇艺PPS -《欢乐颂2》电视剧热播\', is in english:', english_char('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print('String: \'Docs To Go™ Free Office Suite\', is in english:', english_char('Docs To Go™ Free Office Suite'))
print('String: \'Instachat 😜\', is in english:', english_char('Instachat 😜'))

String: 'Instagram', is in english: True
String: '爱奇艺PPS -《欢乐颂2》电视剧热播', is in english: False
String: 'Docs To Go™ Free Office Suite', is in english: True
String: 'Instachat 😜', is in english: True


---

The function is still not perfect, and very few non-English apps might get past our filter, but this seems good enough at this point in our analysis — we shouldn't spend too much time on optimization at this point.

Below, we use the is_english() function to filter out the non-English apps for both data sets:

---

In [9]:
# remove non-english apps for android
android_english = []
for i in android_clean:
    name = i[0]
    english = english_char(name)
    if english:
        android_english.append(i)
        
# remove non-english apps for IOS
ios_english = []
for i in ios_data:
    name = i[1]
    english = english_char(name)
    if english:
        ios_english.append(i)

print('Number of android apps:', len(android_english))
print('Number of IOS apps:', len(ios_english))

Number of android apps: 9614
Number of IOS apps: 6183


In [10]:
# free apps for android apps
android_free_apps = []
for i in android_english:
    price = i[7]
    if price == '0':
        android_free_apps.append(i)

# free apps for IOS apps
ios_free_apps = []
for i in ios_english:
    price = float(i[4])
    if price == 0:
        ios_free_apps.append(i)
        
# assign datasets to new variables
android_final = android_free_apps
ios_final = ios_free_apps

print('Total number of samples (android):', len(android_final))
print('Total number of samples (ios):', len(ios_final))

Total number of samples (android): 8864
Total number of samples (ios): 3222


---

The overall goal is to determine the kinds of apps that are likely to attract more users because the number of people using our apps affect our revenue.

To minimize risks and overhead, our validation strategy for an app idea has three steps:
- Build a minimal Android version of the app, and add it to Google Play.
- If the app has a good response from users, we develop it further.
- If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful in both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification. 

From inspecting the data we can identify the columns that might be useful for determing the most common genres in the market. We'll need to create a frequency table for the 'prime_genre' column in the IOS dataset, and the 'Genres' and 'Category' columns for the android dataset.

---

In [11]:
# function for creating frequency tables for any column
def freq_table(dataset, index, percent=True):
    dictionary = {}
    total = 0
    for i in dataset:
        total += 1
        col = i[index]
        if col in dictionary:
            dictionary[col] += 1
        else:
            dictionary[col] = 1
    
    if percent:
        for i in dictionary:
            percentage = (dictionary[i] / total) * 100
            dictionary[i] = percentage
        
    return dictionary

# function for dispaying a frequency table in decending order
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [12]:
# create frequency table (in percent) for the 'prime_genre' column in the IOS dataset
display_table(ios_final, 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


---


We can see that among the free English apps, more than a half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

---

In [13]:
# create frequency table (in percent) for the 'Category' column in the Android dataset
display_table(android_final, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

---

The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

---

In [14]:
# create frequency table (in percent) for the 'Genres' column in the Android dataset
display_table(android_final, 9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

---

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have the most users.

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the 'Installs' column, but this information is missing for the App Store data set. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the 'rating_count_tot' column.

---

## Most Popular Apps by Genre

---

In [15]:
# compute the average number of user ratings for all genres for IOS apps
ios_genres = freq_table(ios_final, 11)
for genre in ios_genres:
    total = 0
    len_genre = 0
    
    for app in ios_final:
        genre_app = app[11]
        if genre_app == genre:
            n = float(app[5])  # number of ratings for that app
            total += n  # the accumulated total number of ratings for a genre
            len_genre += 1  # the number of apps of that genre
            
    avg = total / len_genre
    print(genre, ':', avg)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


In [16]:
# compute the average number of installs for all genres for Android apps
category_table = freq_table(android_final, 1)

for category in category_table:
    total = 0
    len_category = 0
    
    for app in android_final:
        category_app = app[1]
        if category_app == category:
            n = app[5]
            n = n.replace('+', '')
            n = n.replace(',', '')
            n = float(n)
            total += n
            len_category += 1
    
    avg = total / len_category
    print(category, ':', avg)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

---

## Conclusions

For the IOS apps (Apple Store), we can see that the navigation apps have the highest number of reviews. However, since our system is not entirely accurate, because it doesnt take into account the number of installs, and we have seen previously that entertainment apps have the highest ratings, I would suggest to either make a social networking, navigation or reference app.

For the Android apps (Google Play Store), we can see that game and social categories have the highest total number of installs which I would suggest being the best options for developers to build new apps into. Seeing as more people tend to install those kinds of apps, more people are likely to view and interact with ads which will be more profitable.

Seeing as our stategy is to eventually have the app available in both markets, my overall suggestion would be to develop a an app in the social category as this will likely be successful in both markets.

---